packages/math/src/pool/concentrated/quotes.ts

```diff
@@ -0,0 +1,215 @@
+  amountRemaining: Dec;
+  amountCalculated: Dec;
+  inittedTickIndex: number;
+  sqrtPrice: Dec;
+  liquidity: Dec;
+  feeGrowthGlobal: Dec;
+}
+
+function calcOutGivenIn(
+  tokenIn: Coin,
+  tokenDenom0: string,
+  poolLiquidity: Dec,
+  inittedTicksWithNetLiquidity: TickWithNetLiquidity[],
+  curTick: Int,
+  curSqrtPrice: Dec,
+  precisionFactorAtPriceOne: number,
+  swapFee: Dec,
+  priceLimit = new Dec(0)
+): Int {
+  const isZeroForOne = tokenIn.denom === tokenDenom0;
+  if (isZeroForOne && priceLimit.equals(new Dec(0))) {
+    priceLimit = minSpotPrice;
+  } else if (!isZeroForOne && priceLimit.equals(new Dec(0))) {
+    priceLimit = maxSpotPrice;
+  }
+
+  const sqrtPriceLimit = approxSqrt(priceLimit);
+  const swapStrategy = makeSwapStrategy(isZeroForOne, sqrtPriceLimit, swapFee);
+  const tokenInAmountSpecified = new Dec(tokenIn.amount);
+
+  const curTickWithNetLiq = inittedTicksWithNetLiquidity.find(({ tickIndex }) =>
+    tickIndex.equals(curTick)
+  );
+  if (!curTickWithNetLiq) {
+    throw new Error("curTickNet not found in inittedTicksWithNetLiquidity");
+  }
+  const curTickArrIndex =
+    inittedTicksWithNetLiquidity.indexOf(curTickWithNetLiq);
+  if (curTickArrIndex < 0) {
+    throw new Error(
+      "curTickWithNetLiq not found in inittedTicksWithNetLiquidity"
+    );
+  }
+
+  const swapState: SwapState = {
+    amountRemaining: tokenInAmountSpecified, // tokenIn
+    amountCalculated: new Dec(0), // tokenOut
+    inittedTickIndex: swapStrategy.initTickValue(curTickArrIndex),
+    sqrtPrice: curSqrtPrice,
+    liquidity: poolLiquidity,
+    feeGrowthGlobal: new Dec(0),
+  };
+
+  while (
+    swapState.amountRemaining.gt(smallestDec) &&
+    !swapState.sqrtPrice.equals(sqrtPriceLimit)
+  ) {
+    const nextTick: TickWithNetLiquidity | undefined =
+      inittedTicksWithNetLiquidity?.[swapState.inittedTickIndex];
+    if (!nextTick) {
+      throw new TickOverflowError("Not enough ticks to calculate swap");
+    }
+
+    const nextTickSqrtPrice = tickToSqrtPrice(
+      nextTick.tickIndex,
+      precisionFactorAtPriceOne
+    );
+
+    const sqrtPriceTarget = swapStrategy.getSqrtTargetPrice(nextTickSqrtPrice);
+
+    const {
+      sqrtPriceNext,
+      amountInConsumed: amountOneIn,
+      amountOutComputed: amountZeroOut,
+      feeChargeTotal,
+    } = swapStrategy.computeSwapStepOutGivenIn(
+      swapState.sqrtPrice,
+      sqrtPriceTarget,
+      swapState.liquidity,
+      swapState.amountRemaining
+    );
+
+    swapState.sqrtPrice = sqrtPriceNext;
+    swapState.amountRemaining = swapState.amountRemaining.sub(
+      amountOneIn.add(feeChargeTotal)
+    );
+    swapState.amountCalculated = swapState.amountCalculated.add(amountZeroOut);
+
+    if (nextTickSqrtPrice.equals(sqrtPriceNext)) {
+      const liquidityNet = swapStrategy.setLiquidityDeltaSign(
+        new Dec(nextTick.netLiquidity.toString())
+      );
+
+      swapState.liquidity = addLiquidity(swapState.liquidity, liquidityNet);
+    }
```

Comment:
I suspect that the solution to the other two threads might be just moving the call to `swapStrategy.nextInitializedTickIndex` that is discussed in [this thread](https://github.com/osmosis-labs/osmosis-frontend/pull/1368/files/5863e51afdb574c1d231c4c1bdf5c149e6775207#r1147747087) to be inside this branch because this is the time when we cross the tick.

When we have `!nextTickSqrtPrice.equals(sqrtPriceNext)` (notice nehation), the next initialized tick does not change even though the factual (uninitialized) tick might differ from the previous iteration

Reply:
Done for both swap types

Reply:
Had a follow-up : https://github.com/osmosis-labs/osmosis-frontend/pull/1368/files#r1148474486

but going to resolve this thread to continue discussions in the new one only